Deployment of unpacked WAB fails when using Felix fileinstall

# Description #
----------

If you place an unpacked (unzipped) WAB bundle in a directory managed by Felix's fileinstall for Payara, the deployment fails.  This is because Payara assumes that all URL schemas (except bundle:) have NIO filesystem support, but fileinstall uses its own URL schema (jardir:) for which it does not supply filesystem support.

## Expected Outcome

Such a bundle should deploy correctly as a web application.

## Current Outcome

The deployment fails with the following exception:
```
[2020-01-08T10:45:32.015-0500] [Payara 5.194] [SEVERE] [] [javax.enterprise.web.core] [tid: _ThreadID=157 _ThreadName=pool-23-thread-1] [timeMillis: 1578498332015] [levelValue: 1000] [[
  Container StandardEngine[glassfish-web].StandardHost[server].StandardContext[] has not been started
org.apache.catalina.LifecycleException: java.nio.file.FileSystemNotFoundException: Provider "jardir" not installed
        at org.apache.catalina.core.StandardContext.start(StandardContext.java:5701)
        at com.sun.enterprise.web.WebModule.start(WebModule.java:619)
        at org.apache.catalina.core.ContainerBase.startChildren(ContainerBase.java:1551)
        at org.apache.catalina.core.ContainerBase.start(ContainerBase.java:1223)
        at org.apache.catalina.core.StandardHost.start(StandardHost.java:919)
        at org.apache.catalina.core.ContainerBase.startChildren(ContainerBase.java:1551)
        at org.apache.catalina.core.ContainerBase.start(ContainerBase.java:1223)
        at org.apache.catalina.core.StandardEngine.start(StandardEngine.java:383)
        at org.apache.catalina.startup.Embedded.start(Embedded.java:929)
        at com.sun.enterprise.web.WebContainer.postConstruct(WebContainer.java:631)
        at org.jvnet.hk2.internal.ClazzCreator.postConstructMe(ClazzCreator.java:303)
        at org.jvnet.hk2.internal.ClazzCreator.create(ClazzCreator.java:351)
        at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:463)
        at org.jvnet.hk2.internal.SingletonContext$1.compute(SingletonContext.java:59)
        at org.jvnet.hk2.internal.SingletonContext$1.compute(SingletonContext.java:47)
        at org.glassfish.hk2.utilities.cache.Cache$OriginThreadAwareFuture$1.call(Cache.java:74)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at org.glassfish.hk2.utilities.cache.Cache$OriginThreadAwareFuture.run(Cache.java:131)
        at org.glassfish.hk2.utilities.cache.Cache.compute(Cache.java:176)
        at org.jvnet.hk2.internal.SingletonContext.findOrCreate(SingletonContext.java:98)
        at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2102)
        at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:93)
        at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:67)
        at org.glassfish.internal.data.EngineInfo.getContainer(EngineInfo.java:86)
        at com.sun.enterprise.v3.server.ApplicationLifecycle.startContainers(ApplicationLifecycle.java:1122)
        at com.sun.enterprise.v3.server.ApplicationLifecycle.startEngine(ApplicationLifecycle.java:964)
        at com.sun.enterprise.v3.server.ApplicationLifecycle.setupContainerInfos(ApplicationLifecycle.java:796)
        at com.sun.enterprise.v3.server.ApplicationLifecycle.prepare(ApplicationLifecycle.java:424)
        at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:612)
        at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:606)
        at org.glassfish.osgijavaeebase.OSGiDeploymentRequest.deploy(OSGiDeploymentRequest.java:235)
        at org.glassfish.osgijavaeebase.OSGiDeploymentRequest.execute(OSGiDeploymentRequest.java:149)
        at org.glassfish.osgijavaeebase.AbstractOSGiDeployer.deploy(AbstractOSGiDeployer.java:161)
        at org.glassfish.osgijavaeebase.OSGiContainer.deploy(OSGiContainer.java:206)
        at org.glassfish.osgijavaeebase.JavaEEExtender.deploy(JavaEEExtender.java:159)
        at org.glassfish.osgijavaeebase.JavaEEExtender.access$200(JavaEEExtender.java:49)
        at org.glassfish.osgijavaeebase.JavaEEExtender$HybridBundleTrackerCustomizer$1.call(JavaEEExtender.java:235)
        at org.glassfish.osgijavaeebase.JavaEEExtender$HybridBundleTrackerCustomizer$1.call(JavaEEExtender.java:232)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.nio.file.FileSystemNotFoundException: Provider "jardir" not installed
        at java.nio.file.Paths.get(Paths.java:147)
        at org.glassfish.loader.util.ASClassLoaderUtil.getModulesClasspath(ASClassLoaderUtil.java:267)
        at org.glassfish.loader.util.ASClassLoaderUtil.getModuleClassPath(ASClassLoaderUtil.java:118)
        at com.sun.enterprise.web.WebModuleListener.configureJsp(WebModuleListener.java:259)
        at com.sun.enterprise.web.WebModuleListener.lifecycleEvent(WebModuleListener.java:140)
        at org.apache.catalina.util.LifecycleSupport.fireLifecycleEvent(LifecycleSupport.java:163)
        at org.apache.catalina.core.StandardContext.start(StandardContext.java:5698)
        ... 41 more
]]
```

## Steps to reproduce (Only for bug reports) 

1. Update the osgi.properties to support fileinstall from a separate directory from the autodeploy directory (we use autodeploy/bundles to match the current Glassfish behavior).

2. Install any needed library bundles for the WAB to be deployed (for the Figherfish sample-uas-simplewab sample, install the sample-uas-api bundle) into the fileinstall directory.

3. Unzip a WAB bundle into a subdirectory of the fileinstall directory.  For the above exception we unzipped the Fighterfish sample-uas-simplewab.war into the directory autodeploy/bundles/sample.war.

### Samples

As noted above this can be tested using the sample WAB included with Fighterfish.

## Context

We have a WAB in our application which for various reasons needs to be installed in an unpacked state.  While we recognize this type of auto deployment is generally not recommended for production environments, it provides us a number of benefits we are reluctant to give up. 

I have attached a fairly simple patch that allows this to work correctly: [Payara-jardir.patch.txt](https://github.com/payara/Payara/files/4036130/Payara-jardir.patch.txt)

## Environment ##

- **Payara Version**: 5.194
- **Edition**: Full
- **JDK Version**:  OpenJDK 8
- **Operating System**: Any (tested on Linus and Mac)

--- a/nucleus/deployment/common/src/main/java/org/glassfish/loader/util/ASClassLoaderUtil.java
+++ b/nucleus/deployment/common/src/main/java/org/glassfish/loader/util/ASClassLoaderUtil.java
@@ -263,6 +263,12 @@ public class ASClassLoaderUtil {
                                             "Error truncating URI with prefix of \"reference:\"",
                                             use);
                                 }
+                            } else if (uri.toString().startsWith("jardir:")) {
+                                // OSGi fileinstall created modules can have a "jardir:" schema for directory
+                                // structures, but their is no FileSystemProvider for this type.
+                                // Since the path is the file system path, just substitute the "file:" schema
+                                tmpString.append("file:").append(uri.toString().substring("jardir:".length()));
+                                tmpString.append(File.pathSeparator);
                             } else {
                                 tmpString.append(Paths.get(uri).toString());
                                 tmpString.append(File.pathSeparator);

--- a/nucleus/deployment/common/src/main/java/org/glassfish/loader/util/ASClassLoaderUtil.java
+++ b/nucleus/deployment/common/src/main/java/org/glassfish/loader/util/ASClassLoaderUtil.java
@@ -265,7 +265,7 @@ public class ASClassLoaderUtil {
                                 }
                             } else if (uri.toString().startsWith("jardir:")) {
                                 // OSGi fileinstall created modules can have a "jardir:" schema for directory
-                                // structures, but their is no FileSystemProvider for this type.
+                                // structures, but there is no FileSystemProvider for this type.
                                 // Since the path is the file system path, just substitute the "file:" schema
                                 tmpString.append("file:").append(uri.toString().substring("jardir:".length()));
                                 tmpString.append(File.pathSeparator);

